### PR TITLE
Ctf module

### PR DESCRIPTION
**Pull request by harshit-gupta-cor**
Splitted ctf module out of modules.py as per the [discussion](https://github.com/compSPI/simSPI/pull/20#issuecomment-983853121). 
The comments from the previous [PR](https://github.com/compSPI/simSPI/pull/20) for this module have been incorporated, including, the need for more documentation for the CTF.

### CODE REVIEW

File: simSPI/linear_simulator/ctf_utils.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            envelope = torch.exp(-self.frequency ** 2 * decay ** 2 * self.r2)
+        else:
+            envelope = torch.exp(
+                -self.frequency ** 2 * self.config.b_factor / 4.0 * self.r2
+            )
+
+        hFourier *= envelope
+        return hFourier
+
+    def forward(self, x_fourier, ctf_params={}):
```
Comment (fredericpoitevin) [inline simSPI/linear_simulator/ctf_utils.py:102]: See other PR, please do this and then request reviews again:

```
def forward(self, x_fourier, ctf_params=None):
  if ctf_params is not None:
    ...
```

**Review by fredericpoitevin — CHANGES_REQUESTED** [review]
See comment.

**Comment by geoffwoollard** [comment]
Is this ready for me to review @harshit-gupta-cor ?

File: simSPI/linear_simulator/ctf_utils.py (hunk 1 above)
Comment (harshit-gupta-cor) [reply simSPI/linear_simulator/ctf_utils.py]: Done. Thanks for the suggestion.

**Comment by harshit-gupta-cor** [comment]
@geoffwoollard yes now it is ready to review. I have replied to your comments from the previous PR #20.

**Review by fredericpoitevin — APPROVED** [review]
great CTF doc, @harshit-gupta-cor !!! Thanks for doing this 👍

File: simSPI/linear_simulator/ctf_utils.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    def _get_ewavelength(self):
+        """Output the wavelength of the electron beam.
+
+        Returns
+        -------
+        wavelength: float
+            wavelength of the electron beam in A.
+        """
+        wavelength = 12.2639 / np.sqrt(
+            self.config.kv * 1e3 + 0.97845 * self.config.kv ** 2
```
Comment (geoffwoollard) [inline simSPI/linear_simulator/ctf_utils.py:42]: self.config.kv * (1e3 + 0.97845 * self.config.kv)

**Review by geoffwoollard — APPROVED** [review]
Great documentation!

File: simSPI/linear_simulator/ctf_utils.py (hunk 2 above)
Comment (geoffwoollard) [reply simSPI/linear_simulator/ctf_utils.py:42]: Every little bit counts, right?

Comment (harshit-gupta-cor) [reply simSPI/linear_simulator/ctf_utils.py:42]: Sorry, I think I missed replying to this comment in the previous PR. But I think the current version is more readable, at least for me.

File: simSPI/linear_simulator/ctf_utils.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        """
+        defocus_u = ctf_params["defocus_u"]
+        defocus_v = ctf_params["defocus_v"]
+        angle_astigmatism = ctf_params["defocus_angle"]
+        elliptical = (
+            defocus_v
+            + (defocus_u - defocus_v)
+            * torch.cos(self.angleFrequency - angle_astigmatism) ** 2
+        )
+        defocusContribution = (
```
Comment (ninamiolane) [inline simSPI/linear_simulator/ctf_utils.py:140]: no CamelCase in python, see pep8 https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: tests/test_ctf.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    else:
+        config_dict = {}
+    config = AttrDict(config_dict)
+    return saved_data, config
+
+
+def normalized_mse(a, b):
+    """Return mean square error.
+
+    Calclulates mean square error between two inputs normalized
```
Comment (ninamiolane) [inline tests/test_ctf.py:50]: Typo: ---> `Calculate`  (no s)

File: tests/test_ctf.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    out: torch.Tensor
+        Tensor of size (batch,1, size,size)
+    """
+    r = torch.fft.fftshift(r, dim=(-2, -1))
+    return torch.fft.ifftshift(
+        torch.fft.fftn(r, s=(r.shape[-2], r.shape[-1]), dim=(-2, -1)), dim=(-2, -1)
+    )
+
+
+def test_ctf():
```
[the remark this answers is not in the document]
Comment (ninamiolane) [reply tests/test_ctf.py:85]: `test__get_ewavelength`, `test_get_ctf` and `test_forward`

File: tests/test_ctf.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+"""Test function for CTF module."""
+import numpy as np
+import torch
+
+from simSPI.linear_simulator.ctf_utils import CTF
```
Comment (ninamiolane) [inline tests/test_ctf.py:5]: `ctf_utils` should just be called `ctf.py` (give it the name of the principal object inside it).

**Review by ninamiolane — CHANGES_REQUESTED** [review]
@harshit-gupta-cor @fredericpoitevin @geoffwoollard sorry for the huge delay on my side, I thought I had reviewed this.

Great code! It is refreshing to see this simulation code put into classes.

A few comments:
- all methods of the class `CTF` should be tested: we need one test function per method, even for the private methods.
- other minor remarks are inline.

**Comment by ninamiolane** [comment]
Any update on this?

File: tests/test_ctf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (harshit-gupta-cor) [reply tests/test_ctf.py]: Yeah thanks for pointing out. Done in the new commits.

File: simSPI/linear_simulator/ctf_utils.py (hunk 3 above)
Comment (harshit-gupta-cor) [reply simSPI/linear_simulator/ctf_utils.py]: Sorry missed them. Corrected. Thanks!

File: tests/test_ctf.py (hunk 4 above)
Comment (harshit-gupta-cor) [reply tests/test_ctf.py]: Done. Thanks!

File: tests/test_ctf.py (hunk 6 above)
Comment (harshit-gupta-cor) [reply tests/test_ctf.py]: Done!

**Comment by harshit-gupta-cor** [comment]
> Any update on this?

Yes done!

**Comment by ninamiolane** [comment]
Thanks! Any idea why the tests are failing?

**Comment by harshit-gupta-cor** [comment]
image
Tests associated with TEM simulator are failing (unrelated to this PR). The tests associated with this PR were passing earlier but after the latest PRs to simSPI, the workflow is only testing the test functions already present in the base branch. @fredericpoitevin  is looking into this issue. There were also some issues raised by deepsource which I have now resolved.

**Comment by ninamiolane** [comment]
Excellent, thank you very much for the update! 🙌

**Comment by ninamiolane** [comment]
@harshit-gupta-cor : I think that @fredericpoitevin solved the issue related to the tests! 🙌  

Could you check if this PR can now pass? Many thanks!

**Comment by harshit-gupta-cor** [comment]
I am still getting [errors](https://github.com/compSPI/simSPI/runs/5015771958?check_suite_focus=true). Some problem with checking the linting. @fredericpoitevin  could you please look into it? 
I guess my PR is still pointing to the previous commit of the base branch. Could that be the reason? If yes then is it possible to change the pointer to the latest commit of the base branch without creating a new PR. I couldn't find any solution online.

**Comment by fredericpoitevin** [comment]
Hi @harshit-gupta-cor `codecov/project` is failing because coverage is decreasing compared to `master`.

Looking into the details, I see that testing this line should fix it: https://app.codecov.io/gh/compSPI/simSPI/compare/27/diff#D1L166

Please have a look! Thanks

**Comment by fredericpoitevin** [comment]
The error raised by Deepsource is useful to understand:  https://deepsource.io/gh/compSPI/simSPI/run/769845b7-a1a3-49dd-bee9-970e5b506fd8/python/PYL-W0201

The line of code raising it is:
```python 
config.b_factor = 4 * decay **2
```
and the complaint is that `Defining an instance attribute outside __init__ affects the readability of code.`

What this means is that all attributes of `config` can only be defined at initialization, i.e. in the `__init__` function of a class. Here is is defined in the `AttrDict` class, but then updated at the incriminated line. 

Doing so will lead to error or confusion. Lines 106 and 107 should not be necessary. I have done the following to try and fix this - essentially hard-code what the b-factor value should be in the data test file:
```python 
ctf_data = np.load('tests/data/ctf_data.npy', allow_pickle=True)
config_dict = ctf_data.item()['config_dict']
decay = np.sqrt(-np.log(config_dict['value_nyquist'])) * 2.0 * config_dict['pixel_size'] # equivalent of l.106
ctf_data.item()['config_dict']['b_factor'] = 4 * decay ** 2 # equivalent of l.107
np.save('tests/data/ctf_data.npy', ctf_data, allow_pickle=True)
```
Now `config_dict` read: 
```python
{'side_len': 32, 'batch_size': 4, 'kv': 300, 'pixel_size': 3.2, 'cs': 2.7, 'amplitude_contrast': 0.1, 'ctf_size': 32, 'b_factor': 377.2555416361444, 'value_nyquist': 0.1}
```

**Pull request merged**